modules/core/core-content/src/main/java/com/enonic/xp/core/impl/content/ParentProjectSynchronizer.java

```diff
@@ -0,0 +1,327 @@
+                    parentId( currentContent.getId() ).
+                    recursive( false ).
+                    childOrder( currentContent.getChildOrder() ).
+                    size( -1 ).
+                    build() );
+
+                for ( final Content content : result.getContents() )
+                {
+                    this.doSync( content );
+
+                    if ( content.hasChildren() )
+                    {
+                        queue.offer( content );
+                    }
+                }
+            }
+        } );
+
+        targetContext.runWith( () -> {
+            if ( contentService.contentExists( sourceContent.getId() ) )
+            {
+                cleanDeletedContents( contentService.getById( sourceContent.getId() ) );
+                return;
+            }
+
+            final Content root = sourceContext.callWith( () -> contentService.getByPath( ContentPath.ROOT ) );
+
+            if ( root.getId().equals( sourceContent.getId() ) )
+            {
+                cleanDeletedContents( contentService.getByPath( ContentPath.ROOT ) );
+            }
+        } );
+    }
+
+    private void doSync( final Content sourceContent )
+    {
+        targetContext.runWith( () -> {
+            final Content targetContent =
+                contentService.contentExists( sourceContent.getId() ) ? contentService.getById( sourceContent.getId() ) : null;
+
+            final ProjectEventSyncCommandParams commandParams = createEventCommandParams( sourceContent, targetContent );
+
+            final List<AbstractProjectEventSyncCommand> commands = createEventCommands( commandParams, targetContent != null ? EnumSet.of(
+                ProjectSyncEventType.MOVED, ProjectSyncEventType.RENAMED, ProjectSyncEventType.SORTED,
+                ProjectSyncEventType.MANUAL_ORDER_UPDATED, ProjectSyncEventType.UPDATED ) : EnumSet.of( ProjectSyncEventType.CREATED ) );
+
+            commands.forEach( AbstractProjectEventSyncCommand::sync );
+        } );
+    }
+
+    private ProjectEventSyncCommandParams createEventCommandParams( final Content sourceContent, final Content targetContent )
+    {
+        return ProjectEventSyncCommandParams.create().
+            sourceContext( sourceContext ).
+            targetContext( targetContext ).
+            sourceContent( sourceContent ).
+            targetContent( targetContent ).
+            build();
+    }
+
+    private List<AbstractProjectEventSyncCommand> createEventCommands( final ProjectEventSyncCommandParams params,
+                                                                       final EnumSet<ProjectSyncEventType> syncTypes )
+    {
+        final ImmutableList.Builder<AbstractProjectEventSyncCommand> commands = ImmutableList.builder();
+
+        if ( syncTypes.contains( ProjectSyncEventType.CREATED ) )
+        {
+            if ( params.getSourceContent() != null )
```

Comment:
These ifs look suspicions. Don't we expect `getSourceContent` getTargetContent` to be non-null for certain events?

Reply:
no, getSourceContent() is null for `deleted` event, getTargetContent() - for `created`